From fb48900977dbbfc6f921822abc6728b6075efc4f Mon Sep 17 00:00:00 2001
From: Anton Gerasimov <anton.gerasimov@here.com>
Date: Fri, 1 Feb 2019 14:39:48 +0100
Subject: [PATCH] board: raspberrypi: add serial and revision to device tree

Raspberry Pi bootloader adds this node to fdt, but if u-boot script
doesn't reuse the tree provided by it, this information is lost.

Revision and serial are displayed in /proc/cpuinfo after boot.

Signed-off-by: Anton Gerasimov <anton.gerasimov@here.com>
Signed-off-by: Patrick Vacek <patrick.vacek@here.com>

---
 board/raspberrypi/rpi/rpi.c | 31 +++++++++++++++++++++++++++++--
 1 file changed, 29 insertions(+), 2 deletions(-)

diff --git a/board/raspberrypi/rpi/rpi.c b/board/raspberrypi/rpi/rpi.c
index e367ba3092..d70e07f0a6 100644
--- a/board/raspberrypi/rpi/rpi.c
+++ b/board/raspberrypi/rpi/rpi.c
@@ -252,6 +252,8 @@ static uint32_t rev_scheme;
 static uint32_t rev_type;
 static const struct rpi_model *model;
 
+uint64_t serial;
+
 int dram_init(void)
 {
 	ALLOC_CACHE_ALIGN_BUFFER(struct msg_get_arm_mem, msg, 1);
@@ -384,8 +386,8 @@ static void set_serial_number(void)
 		return;
 	}
 
-	snprintf(serial_string, sizeof(serial_string), "%016llx",
-		 msg->get_board_serial.body.resp.serial);
+	serial = msg->get_board_serial.body.resp.serial;
+	snprintf(serial_string, sizeof(serial_string), "%016llx", serial);
 	env_set("serial#", serial_string);
 }
 
@@ -478,6 +480,29 @@ void *board_fdt_blob_setup(void)
 	return (void *)fw_dtb_pointer;
 }
 
+static int ft_add_revision_info(void *blob) {
+	int off;
+	int ret;
+
+	off = fdt_subnode_offset(blob, 0, "system");
+
+	if (off < 0) {
+		off = fdt_add_subnode(blob, 0, "system");
+		if (off < 0)
+			return -1;
+	}
+
+	ret = fdt_setprop_u64(blob, off, "linux,serial", serial);
+	if (ret < 0)
+		return -1;
+
+	ret = fdt_setprop_u32(blob, off, "linux,revision", revision);
+	if (ret < 0)
+		return -1;
+
+	return 0;
+}
+
 int ft_board_setup(void *blob, bd_t *bd)
 {
 	/*
@@ -487,6 +512,8 @@ int ft_board_setup(void *blob, bd_t *bd)
 	 */
 	lcd_dt_simplefb_add_node(blob);
 
+	ft_add_revision_info(blob);
+
 #ifdef CONFIG_EFI_LOADER
 	/* Reserve the spin table */
 	efi_add_memory_map(0, 1, EFI_RESERVED_MEMORY_TYPE, 0);
